File: Topological Sort/FindEventualSafeStates.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class FindEventualSafeStates {
    public static void main(String[] args) {
        FindEventualSafeStates solution = new FindEventualSafeStates();
        int[][] graph = { { 1, 2 }, { 2, 3 }, { 5 }, { 0 }, { 5 }, {}, {} };
        List<Integer> ans = solution.input(graph);
        System.out.println(ans);
    }

    public List<Integer> input(int[][] graph) {
        int n = graph.length;
        int[] outdegree = new int[n];
        List<List<Integer>> connected = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            connected.add(new ArrayList<>());
        }
        List<Integer> ans = new ArrayList<>();
        Queue<Integer> q = new LinkedList<>();
        for (int i = 0; i < n; i++) {
            outdegree[i] = graph[i].length;
            if (outdegree[i] == 0) {
                ans.add(i);
                q.add(i);
            }
        }
        for (int i = 0; i < n; i++) {
            for (int e : graph[i]) {
                connected.get(e).add(i);
            }
        }
        while (!q.isEmpty()) {
            int node = q.poll();
            for (int e : connected.get(node)) {
                outdegree[e]--;
                if (outdegree[e] == 0) {
                    ans.add(e);
                    q.add(e);
                }
            }
        }
        Collections.sort(ans);
        return ans;
        // int n = graph.length;
        // int[] outdegree = new int[n];
        // List<List<Integer>> connected = new ArrayList<>();
        // for (int i = 0; i < n; i++) {
        // connected.add(new ArrayList<>());
        // }
        // List<Integer> ans = new ArrayList<>();
        // Queue<Integer> q = new LinkedList<>();
        // for (int i = 0; i < n; i++) {
        // outdegree[i] = graph[i].length;
        // if (outdegree[i] == 0) {
        // ans.add(i);
        // q.add(i);
        // }
        // }
        // for (int i = 0; i < n; i++) {
        // for (int e : graph[i]) {
        // connected.get(e).add(i);
        // }
        // }
        // while (!q.isEmpty()) {
        // int node = q.poll();
        // for (int e : connected.get(node)) {
        // outdegree[e]--;
        // if (outdegree[e] == 0) {
        // ans.add(e);
        // q.add(e);
        // }
        // }
        // }
        // Collections.sort(ans);
        // return ans;
    }
}

File: Topological Sort/FindEventualSafeStates.js
var FindEventualSafeStates=function(graph){
    let n=graph.length;
    let revGraph={};
    let inDegree=new Array(n).fill(0)
    for(let i=0;i<n;i++){
        for(let x of graph[i]){
            if(!revGraph[x]){
                revGraph[x]=[]
            }
            revGraph[x].push(i)
            inDegree[i]++
        }
    }

    let queue=[]
    let result=[]
    for(let i=0;i<n;i++){
        if(inDegree[i]==0){
            queue.push(i)
        }
    }
    while(queue.length){
        let current=queue.shift()
        result.push(current)
        let neighour=revGraph[current]
        if(neighour){
            for(let x of neighour){
                inDegree[x]--
                if(inDegree[x]==0){
                    queue.push(x)
                }
            }
        }
    }
    return result.sort((a,b)=>a-b)
}
const graph=[[1,2],[2,3],[5],[0],[5],[],[]];
const ans=FindEventualSafeStates(graph);
console.log(ans);